File: array_and_strings/sliding_window/c++/min_subarray_len/min_subarray_len.cpp
// min_subarray_len.cpp
#include "min_subarray_len.h"
#include <vector>
#include <limits.h>

int minSubarrayLen(const std::vector<int> &nums, int target)
{
    int currSum = 0, left = 0;
    int windowLen = INT_MAX;

    for (int right = 0; right < nums.size(); right++)
    {
        // Add the next element from the right
        currSum += nums[right];
        // Once we achieve currSum >= target, we can no longer find another valid, smaller window anymore, need to shrink
        while (currSum >= target)
        {
            // Check if the current window size is smaller than the previous one
            windowLen = std::min(windowLen, right - left + 1);
            // Shrink the window from the left
            currSum -= nums[left];
            left++;
        }
    }

    // Return 0 if no such subarray is found, and the window length remains INT_MAX
    return (windowLen != INT_MAX) ? windowLen : 0;
}

File: array_and_strings/sliding_window/python3/min_subarray_len.py
from math import inf, isinf
from typing import List


def min_subarray_len(nums: List[int], target: int) -> int:
    """
    Given an array of positive integers `nums` and a positive integer `target`, return the minimal length of a subarray whose sum is greater 
    than or equal to target. If there is no such subarray, return 0 instead.
    
    Parameters
    ----------
    nums : List[int]
        List of positive integers
    target : int
        Positive integer
        
    Returns
    -------
    int
        Minimal length of a subarray whose sum is greater than or equal to target
    """
    left = curr_sum = 0 
    window_len = inf
    for right in range(len(nums)):
        # Add elements from the right (expand window)
        curr_sum += nums[right]
        # Once we achieve window sum >= target, we can no longer do any better (i.e find a smaller valid subarray) with the left index at the current position
        # This loop stops running when window sum has become < target again
        while (curr_sum >= target):
            # Update if the current window size is smaller than the previously found window size 
            window_len = min(window_len, right - left + 1)
            # Remove the element pointed to by the left pointer and increment the left index to shrink the window
            # Shrink only after we have checked if window_len could be updated
            curr_sum -= nums[left]
            left += 1
      
    return window_len if not isinf(window_len) else 0


def main() ->  int:

    nums = [2, 3, 1, 2, 4, 3]
    target = 7
    
    result = min_subarray_len(nums, target)
    
    print(f"Given the input {nums} and target = {target}, the smallest subarray with a sum of >= {target} has length {result}")
    
    return 0

if __name__ == "__main__":
    
    main()
